core/src/oauth/store.rs

```diff
@@ -0,0 +1,173 @@
+use crate::oauth::connection::{Connection, ConnectionProvider, ConnectionStatus};
+use crate::utils;
+use anyhow::Result;
+use async_trait::async_trait;
+use bb8::Pool;
+use bb8_postgres::PostgresConnectionManager;
+use tokio_postgres::NoTls;
+
+#[async_trait]
+pub trait OAuthStore {
+    async fn create_connection(
+        &self,
+        provider: ConnectionProvider,
+        metadata: serde_json::Value,
+    ) -> Result<Connection>;
+    async fn retrieve_connection(
+        &self,
+        provider: ConnectionProvider,
+        connection_id: &str,
+    ) -> Result<Connection>;
+
+    fn clone_box(&self) -> Box<dyn OAuthStore + Sync + Send>;
+}
+
+impl Clone for Box<dyn OAuthStore + Sync + Send> {
+    fn clone(&self) -> Self {
+        self.clone_box()
+    }
+}
+
+#[derive(Clone)]
+pub struct PostgresOAuthStore {
+    pool: Pool<PostgresConnectionManager<NoTls>>,
+}
+
+impl PostgresOAuthStore {
+    pub async fn new(db_uri: &str) -> Result<Self> {
+        let manager = PostgresConnectionManager::new_from_stringlike(db_uri, NoTls)?;
+        let pool = Pool::builder().max_size(16).build(manager).await?;
+        Ok(Self { pool })
+    }
+
+    pub async fn init(&self) -> Result<()> {
+        let conn = self.pool.get().await?;
+        for table in POSTGRES_TABLES {
+            conn.execute(table, &[]).await?;
+        }
+        for index in SQL_INDEXES {
+            conn.execute(index, &[]).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl OAuthStore for PostgresOAuthStore {
+    async fn create_connection(
+        &self,
+        provider: ConnectionProvider,
+        metadata: serde_json::Value,
+    ) -> Result<Connection> {
+        let pool = self.pool.clone();
+        let c = pool.get().await?;
+
+        let created = utils::now();
+        let status = ConnectionStatus::Pending;
+        let secret = utils::new_id();
+
+        // Create connection
+        let stmt = c
+            .prepare(
+                "INSERT INTO connections (id, created, provider, secret, status, metadata)
+                   VALUES (DEFAULT, $1, $2, $3, $4, $5::jsonb) RETURNING id",
+            )
+            .await?;
+        let row_id: i64 = c
+            .query_one(
+                &stmt,
+                &[
+                    &(created as i64),
+                    &serde_json::to_string(&provider)?,
```

Comment:
Can't we implement toStr on the Provider so we can also log it?

Reply:
Sure

Reply:
(done, also FromStr)

Reply:
👌